Insights Operator: Gather Workload Runtime Info From Containers (#1584)
openshift/enhancements: enhancements/insights/insights-runtime-extractor.md

### API Extensions

This proposal does not require any API Extensions to the OpenShift Insights Operator.

### Topology Considerations

#### Hypershift / Hosted Control Planes

#### Hypershift [optional]

This proposal only aims to gather data about the Data plane and will not inspect any containers running in the Control Plane.

#### Standalone Clusters

This proposal is relevant for standalone clusters.

#### Single-node Deployments or MicroShift

This proposal affects single-node deployment as there is additional resource consumption
when the Insights Operator gathers workload data that will affect the only worker node.

### Implementation Details/Notes/Constraints [optional]

#### Container Scanner API

The Insights Operator communicates with the insights-runtime-extractor pods deployed by its DaemonSet with a simple HTTP server that has the following API:

* `GET /gather-runtime-info`
  * Inspect the all the containers running on the worker node. The HTTP response returns a JSON payload with the extracted data organized in a tree of `namespace/pod-name/container-id/runtime-info`:

```json
{
  "my-namespace": {

>     "my-pod-1": {

Reviewer: When I look at your integration work now, I wonder if we need this structure. Are the namespace and Pod names needed? 

Another question is whether we could use `imageID` as a key for the below structure? I guess it could help avoid some duplication (especially on the "aggregation" level in the Insights Operator). - e.g when I deploy my app with 3 replicas then I will have three various containerIDs (on various nodes probably) comparing to the same `imageID`

Author: I don't think `imageID` is a good key to identify workloads. Eg you could have the same `fedora` image that runs very different workloads based on the `args` (or environment variables)

Any code running under such a high level of privilege is an attractive target for privilege escalation, since it can be leveraged to gain complete access to the node’s underlying operating system, and all hosted data. It is therefore essential to minimize the potential attack surface area and harden all essential code of the insights-runtime-extractor component.

##### Reducing Attack Surface Area through Modular Process Design

> Splitting the inspection logic into multiple plugins/modules not only allows for extensibility but also provides for the ability to partition critical privilege requiring code from code which requires only point in time access to a specific container. Following this approach, privileged actions can be limited to a thin coordinator process, with the bulk of logic executed as a forked downgraded process limited to the same access afforded the inspected container target. Additionally, limiting the lifespan of the downgraded logic’s process to a per-container execution basis mitigates potential attacks where a rogue container corrupts the inspection logic to modify or gain access to a second victim container. 

Reviewer 2: Nit: This sections refer to `coordinator` process, while the word is not mentioned in any of the text above. Would be good to keep the terminology consistent to reduce confusion.

As such, three key key rules should be followed in the design and implementation of the coordinator process

1. __Dynamic loading of arbitrary executable code should never be supported or allowed by the coordinator process.__ This precludes scripting language runtimes (such as python and JS) as well as bytecode interpreting languages such as Java, .Net, and WASM. It also precludes usage of dlopen(), or libraries with dynamic loading capabilities: the process must be statically compiled. 
1. __Dependencies in the coordinator should be kept to only the provided language SDK, and Linux system contracts__, even at the expense of some code duplication
1. __Coordinator code should be limited to what is strictly required__ to execute in a privileged manner. All non-privilege requiring code should be externalized into the logic of other processes.

##### Defensive Input and Memory Guarding

> While keeping the bulk of analysis logic out of the coordinator process reduces risk, there is still the potential for a deprivileged insights-runtime-extractor component to influence the coordinator as a corrupted relay through its required communication channel. As an example, a corrupted container could lead a insights-runtime-extractor process to write a carefully crafted payload which triggers a buffer overflow and potentially arbitrary code execution in the coordinator. Therefore the coordinator process should take particular care in the handling of all input, assuming potentially hostile data. Additionally, the coordinator process should take advantage of technologies and techniques that mitigate memory corruption defects. As an example, Rust’s memory access model provides strong guarantees of memory safety, and limits state mutability to the smaller set of intended write code paths. Usage of a defensive language like Rust over a traditional memory unsafe language like C would mitigate risk further by providing a second line of defense, as well as preventing a number of common issues frequently leveraged in exploits. 

Reviewer 2: You can update this to reflect on the idea of the coordinator to not read the output of the fingerprint executables.

---
title: insights-runtime-extractor
authors:
  - "@jmesnil"
reviewers:
  - "@tremes"
  - "@iNecas"
approvers:
  - "@mrunalp"
api-approvers:
  - TBD (not relevant as this RFE does not impact API?)
creation-date: 2024-03-04
last-updated: 2024-07-23
status: implementable
tracking-link: 
  - https://issues.redhat.com/browse/OBSDA-850
see-also:
  - "https://issues.redhat.com/browse/CCXDEV-11505"
  - "https://issues.redhat.com/browse/CCX-138"
  - "https://issues.redhat.com/browse/CCXDEV-12963"
---

# Gather Workload Runtime Info From Containers

This document proposes a mechanism to gather runtime info from containers to get additional insights on user workloads and make investment decisions that will drive utilization.

## Summary

The Insights operator periodically gathers cluster data and sends it to Red Hat. The data about workloads consists of the image ID and the entrypoint. 

> In many cases, the image ID and its layers do not provide any meaningful information about the actual application or application language/platform/framework...  

Reviewer 3: ```suggestion
In many cases, the image ID and its layers do not provide any meaningful information about the actual application or application language/platform/framework.
```

To get this level of information, we can gather additional data on the running containers.   

> By inspecting running containers, the Insights Operator can enrich the gathered data and capture the kind of workload that is running on OpenShift. This information can be used by Red Hat to target  OpenShift investments at the most popular workloads, inform decisions around long-life offerings for application platforms and frameworks, identify growth opportunities in emerging workloads, and measure the success of our strategic initiatives in AI and Middleware.

Reviewer 3: ```suggestion
By inspecting running containers, the Insights Operator can enrich the gathered data and capture the kind of workload that is running on OpenShift. This information can be used by Red Hat to target OpenShift investments at the most popular workloads, inform decisions around long-life offerings for application platforms and frameworks, identify growth opportunities in emerging workloads, and measure the success of our strategic initiatives in AI and Middleware.
```

## Motivation

Currently, the Insights operator tries to identify what workload is used on OpenShift by looking for known fingerprinting in layers of containers, but there are a lot of issues with this approach where we can only identify base-level containers.
We do, for example, not know if a python container is using AI libraries.
Another issue is that we can only identify containers using known base containers (like OpenJDK), but we do not know what type of workload is more popular (Spring Boot, Quarkus, WildFly, EAP, etc.).

The Insights operator can also not tell if a container is using a non-Red Hat provided base containers like Ubuntu or Debian. Knowing this data means that BU and product leadership can use this data to prioritize future work and realign investment.

The proof of concept for this has been demonstrated to Red Hat leeadership, who agree that this is important data for us to know in order 
to help shape the future of our product offerings.

### User Stories

* As a middleware product manager, I can observe the trends of a product version over time to estimate the expected value of any customer requested extended life offering.

> * As a product manager in the AI space, I can measure the application platforms/frameworks/languages used in clusters or namespaces running OpenShift AI to understand what users are coding applications in.

Reviewer 3: so we could offer up-to-date images with those libraries and think how about how to support those better.

* As a middleware product manager, I can track the number of applications built using Quarkus to understand our return on investment from an OpenShift workload perspective.

* As a RHEL product manager, I can follow trends in the use of other Linux containers to inform my product roadmap.

> * As an OpenShift product marketing manager, I want to be able to produce content that shows users use OpenShift as a  platform for all types of workloads.

Reviewer 3: ```suggestion
* As an OpenShift product marketing manager, I want to be able to produce content that shows users use OpenShift as a platform for all types of workloads.
```

    

### Goals

> The goal of this proposal is to inspect running containers and pop a few key pieces of data to identify the workload that they are running. This data will enrich and complement the information already gathered by the Insights Operator to get a more accurate representation of workloads running on OpenShift.

Reviewer 3: nit: pop --> surface?

### Non-Goals

> * It is not a goal of this proposal to scan container for security purposes

Reviewer 3: ```suggestion
* It is not a goal of this proposal to scan containers for security purposes
```

From the operator perspective, the insights-runtime-extractor is a black box, and its requirements are:

* Determine the Pod name of each insights-runtime-extractor’s Pod (as the insights-runtime-extractor is deployed as a DaemonSet, there is a single pod instance on each worker node) and map it to its worker node.
  * There is no service exposing the insights-runtime-extractor and its pods are accessed directly throug their DNS records.
* When the operator starts gathering workload information, query the `HTTP GET /gather-runtime-info` request on each insights-runtime-extractor pod. The HTTP response will return a JSON payload in the standard output that represents a tree of `namespace/pod-name/container-id/runtime-info`.
* As the operator iterates over the containers in the cluster, check if there is a runtime-info entry for the `namespace/pod-name/container-id` and add it to the existing `workloadContainerShape` (after hashing all the values)

The insights-runtime-extractor functionality is provided by an image that must be pullable from the `openshift-insights` namespace and present in the cluster at installation.

#### Variation and form factor considerations [optional]

> This enhancement to the Insight Operator is not impacting the Control Plane of OpenShift and should be available regardless of the  form of OpenShift (Standalone OCP,  Microshift, Hypershift).

Reviewer 3: ```suggestion
This enhancement to the Insight Operator is not impacting the Control Plane of OpenShift and should be available regardless of the form of OpenShift (Standalone OCP,  Microshift, Hypershift).
```

> This enhancement only gather runtime info from Linux worker nodes and will not touch Windows worker nodes. The DaemonSet will be configured to only be deployed on Linux worker nodes (having the label `beta.kubernetes.io/os=linux`).

Reviewer 3: ```suggestion
This enhancement only gathers runtime information from Linux worker nodes and will not touch Windows worker nodes. The DaemonSet will be configured to only be deployed on Linux worker nodes (having the label `beta.kubernetes.io/os=linux`).
```

  * Inspect the all the containers running on the worker node. The HTTP response returns a JSON payload with the extracted data organized in a tree of `namespace/pod-name/container-id/runtime-info`:

```json
{
  "my-namespace": {
    "my-pod-1": {
      "cri-o://xxxx": {
        "os-release-id": "rhel",
        "os-release-version-id": "8.8",
        "runtime-kind": "Java",
        "runtime-kind-version": "17.0.8",
        "runtime-kind-implementer": "Red Hat, Inc.",
        "runtimes": [
          {
            "name": "Quarkus",
            "version": "3.4.1",
          }
        ],
      },
      "cri-o://yyyy": {
        "os-release-id": "rhel",
        "os-release-version-id": "9.2",
      }
    },
    "my-pod-2": {
      [...]
    }
  },
  "other-namespace": {
    "other-pod-1": {
      "cri-o://zzzz": {
        [...]
      }
    }
  }
}
```

#### insights-runtime-extractor Implementation

Inspecting a container is achieved by running on the same worker node that the inspected container and looking at the host process table. The insights-runtime-extractor enters the container’s process namespaces to extract meaningful information about the runtime stack.

> The extraction of data from the running container is achieved by executing “fingerprint” executables in the process `mnt` namespace to find the relevant information by reading specific files (e.g. the `Os` and `OsVersion` values are read from the `/etc/os-release` file if it is present). These fingerprints executables are self-contained executables

Reviewer 3: ```suggestion
The extraction of data from the running container is achieved by executing “fingerprint” executables in the process `mnt` namespace to find the relevant information by reading specific files (e.g. the `Os` and `OsVersion` values are read from the `/etc/os-release` file if it is present). These fingerprints executables are self-contained executables.
```

This cannot be achieved by a program running in the OpenShift Insights Operator deployment.
This requires an additional resource for the operator, which is a DaemonSet that would deploy a “insights-runtime-extractor” image on all the cluster worker nodes.

The insights-runtime-extractor image requires high privileges to run (access the worker node’s host table, enter process namespaces, etc.) and needs to be configured with a specific security context constraint.

To minimize the code path running with high privileges, the insights-runtime-extractor is split into 2 components:

> * an "extractor" running in a privileged container. It runs a simple TCP server that, when  it is trigged, will extract the runtime information and store them in a shared volume. The TCP server is bould to the loopback address so that it can only be contacted by clients running inside the same pod.

Reviewer 3: ```suggestion
* an "extractor" running in a privileged container. It runs a simple TCP server that, when it is trigged, will extract the runtime information and store them in a shared volume. The TCP server is bould to the loopback address so that it can only be contacted by clients running inside the same pod.
```

IRE->>V: create a /data/out-${timestamp} dir to collect all fingerprints output
IRE->>IRE: call /crictl ps to list all containers runnnin on the worker node<br>and get their pod name and namespaces
loop For each running container
    IRE->>IRE: Call /crictl inspect to get the root PID of the container
    IRE->>IRE: Get process info from /proc (name, cmd-line, env)
    IRE->>V: Store the process fingerprint
    create participant C as child process
    IRE->>C: Fork the extractor process
    C->>C: Enter namespaces (mnt, pid) of the container process
    C->>C: Switch user
    Note over C: executing inside the namespaces<br>of the running container process
    C->>C: Find all all fingerprints executables that<br> can be applied to the process
    loop For each applied fingerprint
       C->>V: Execute the fingerprint and store its output in the timestamp dir
    end
    destroy C
    IRE-xC: Wait until the child process is finished...
end
IRE->>EXP: Return the path of the timestamp dir
IRE-xEXP: Close the TCP connection
```

#### insights-runtime-extractor Constraints

The insights-runtime-extractor `extractor` is sensitive software that requires high privileges to function.

> It needs to run as privileged and access the host table of Worker Nodes to read their `/proc` tables and run as root to be able to enter in process namespaces.

Reviewer 3: It would need to run in the host PID namespace.

A specific `SecurityContextConstraints` should be added to the `openshift-insights` to deploy the insights-runtime-extractor with the required permissions.

In order to minimize any security risks, its code will be audited for security and it image will be as minimal as possible.

The `extractor` image contains:

* The TCP server to trigger runtime info extraction (written in Rust)

>   * Access to this server is constrained by liseting only to the loopback address. Only clients in the same pod will be able to connect to it.

Reviewer 3: ```suggestion
  * Access to this server is constrained by listening only to the loopback address. Only clients in the same pod will be able to connect to it.
```

*  The `crictl` tool to find the running containers and their root PID
* “Fingerprints” which are self-contained executables that run in the container’s process namespaces to identify what the container is running.

> 

Reviewer 3: nit: drop empty lines

The `exporter` component runs an unprivileged HTTP server. When it receives a `GET /gather-runtime-info`, it triggers an extraction by connecting to the `extractor` that replies
with a path to a dir on the shared volume. The exporter then collects the raw data from that directory, bundle them in a JSON payload sent as the HTTP response.
Connections to the `exporter` HTTP server will be fronted by `kube-rbac-proxy` so that only the Insights Operator is allowed to query it.

The `exporter` image would contain:

* The HTTP server (written in Go)

> Due to the self-contained nature of these executables, the 2 images that compose the insights-runtime-extractor are built from `scratch` (and not from an existing base image).

Reviewer 3: May need to use a minimal ubi image for FIPS.

Author: Thanks, I've amended the RFE. I put the FIPS compliance requirement in the `GA` stage but if you think that this must done for `TechPreview` please let me know.

2. __All analysis I/O operations should be limit-bound.__ A container either through error or design may have files of arbitrary size, or files in a corrupt state. Any potentially unbounded operation (e.g. reading to the end of a file or following a reference data structure to its arbitrary end) is prohibited.
3. __Memory and CPU usage should not be input derived.__ In addition to achieving system stability and performance goals, ensuring the overhead of a inspecting process does not increase significantly relative to input data is critical in preventing potential Denial-of-Service attacks relying on a specially crafted container filesystem.  

##### No Impact on Workloads

Inspecting containers must not disrupt the containers that are inspected and have an impact on their quality of service.
The insights-runtime-extractor is running with its own memory and CPU requirements and does not consume resources from the inspected container.

###### Risk of misbehaving code execution

> A few fingerprints relies on executing process in the user namespace (for example, executing `node --version` to determine the version of Node.js).

Reviewer 3: Starting a lot of processes at the same time for all pods could lead to node instability if node is already near limits. Is there any plan to spread out the scan over some period of time?

Author: I updated the `No Impact on Workloads` section above to clarify the behaviour:

> If the worker node is near limits, a burst of activity from the insights runtime extractor could lead to node instability. To prevent this, the insights runtime extractor will extract runtime information from processes one at a time and will not parallelize these tasks.
 Increasing the gathering time is a trade-off to prevent node instability.

Does that answer your question?

The additional data in the `workloadRuntimeInfoContainer` will increase the size of the insights payload that is archived and sent to Red Hat backends.

The additional data adds __~200 uncompressed bytes per WorkloadContainerShape in the JSON payload__.

This additional payload does not scale with the number of pods but with the number of `WorkloadContainerShape` which is difficult to estimate. There is an upper limit of 8000 pods
in the Insights operator but that does not give a limit on the number of containers. It is also very likely that some of the pods are replicas and their containers would only count for 1 `WorkloadContainerShape` in their gathered data.

#### UX Risk & Mitigation

The additional runtime info added to the gathered data will be exposed to the user though the archives generated by the Insights operator. As the values of the field are hashed, the user will not be able to determine their corresponding content.

> Exposing these data to the user with a clean user interface (eg within console.redhat.com) is out of scope of this proposal.

Reviewer 3: Maybe this should be considered in the future as a cluster wide view of pods running will be very useful to understand how a cluster is utilized.

### Drawbacks

> If implemented, Red Hat will document the addition of this new collection. We  introduce the risk that some number of customers who would have or are sending telemetry data via the Insights operator will choose to turn it off because they are uncomfortable sharing this additional data. 

Reviewer 3: Is there a knob to just control whether this data is collected or not?

Author: No. I added a note about this as something that could be addressed in the `GA` stage by adding a flag in the openshift insights operator configuration.

### Non-Goals

* It is not a goal of this proposal to scan containers for security purposes
* It is not a goal of this proposal to get runtime information specific to a given customer or application
* It is not a goal of this proposal to capture personal sensitive information

## Proposal

This enhancement proposal introduces additional data to identify the technology stack of running containers.

The existing `workloadContainerShape` type that contains workload information about a container will be enhanced with a new `workloadRuntimeInfoContainer` field:

```go
// workloadContainerShape describes the shape of a container which includes
// a subset of the data in the container.
// TODO: this may desirable to make more precise with a whole container hash
//
//  that includes more of the workload, but that would only be necessary if
//  it assisted reconstruction of type of workloads.
type workloadContainerShape struct {
   // ImageID is the content addressable hash of the image as observed from
   // the status or the spec tag.
   ImageID string `json:"imageID"`
   // FirstCommand is a hash of the first value in the command array, if
   // any was set.

>    FirstCommand string `json:"firstCommand,omitempty"`

Reviewer 3: nit: This could be called just `Command` ?

Author: It is an existing field in this struct.
This RFE only adds the new `workloadRuntimeInfoContainer` field.

   // FirstArg is a hash of the first value in the arguments array, if any
   // was set.

>    FirstArg string `json:"firstArg,omitempty"`

Reviewer 3: And this  `Args`?

Author: idem

### Implementation Details/Notes/Constraints [optional]

#### Container Scanner API

The Insights Operator communicates with the insights-runtime-extractor pods deployed by its DaemonSet with a simple HTTP server that has the following API:

* `GET /gather-runtime-info`
  * Inspect the all the containers running on the worker node. The HTTP response returns a JSON payload with the extracted data organized in a tree of `namespace/pod-name/container-id/runtime-info`:

```json
{
  "my-namespace": {
    "my-pod-1": {
      "cri-o://xxxx": {
        "os-release-id": "rhel",
        "os-release-version-id": "8.8",
        "runtime-kind": "Java",
        "runtime-kind-version": "17.0.8",
        "runtime-kind-implementer": "Red Hat, Inc.",
        "runtimes": [
          {
            "name": "Quarkus",
            "version": "3.4.1",
          }
        ],
      },
      "cri-o://yyyy": {
        "os-release-id": "rhel",
        "os-release-version-id": "9.2",
      }
    },
    "my-pod-2": {
      [...]
    }
  },
  "other-namespace": {
    "other-pod-1": {
      "cri-o://zzzz": {
        [...]
      }
    }
  }
}
```

#### insights-runtime-extractor Implementation

> Inspecting a container is achieved by running on the same worker node that the inspected container and looking at the host process table. The insights-runtime-extractor enters the container’s process namespaces to extract meaningful information about the runtime stack.

Reviewer 3: Can we clarify here that it will just join the namespaces and *not* the cgroups so it doesn't count towards container limits?

Author: done in https://github.com/openshift/enhancements/blob/e083208a087afcc9745bd0c0e154174985e16c76/enhancements/insights/insights-runtime-extractor.md#insights-runtime-extractor-constraints

##### No Impact on Workloads

Inspecting containers must not disrupt the containers that are inspected and have an impact on their quality of service.
The insights-runtime-extractor is running with its own memory and CPU requirements and does not consume resources from the inspected container.

If the worker node is near limits, a burst of activity from the insights runtime extractor could lead to node instability. To prevent this,
the insights runtime extractor will extract runtime information from processes one at a time and will not parallelize these tasks.
Increasing the gathering time is a trade-off to prevent node instability.

###### Risk of misbehaving code execution

A few fingerprints relies on executing process in the user namespace (for example, executing `node --version` to determine the version of Node.js).
There is a risk of side-effects if this execution does not match the expectation (if the `node` process is not the Node.js executable and its `--version` parameter triggers a `rm -rf /`).

> Executing such code should be done conservatively on well-identified workloads (`python`, `node`, etc.) if they provide a way to get their versions without side-effect.

Reviewer 3: Another issue here is that when we exec processes inside a container, we are using up the container's resources. If it is already near its memory limit then it could get OOMKilled by the kernel.

Reviewer 3: I see that we mention joining namespaces of the process. Let's clarify that we won't join the cgroups.

Author: I mention that we don't enter the `cgroups` in https://github.com/openshift/enhancements/blob/e083208a087afcc9745bd0c0e154174985e16c76/enhancements/insights/insights-runtime-extractor.md#insights-runtime-extractor-constraints.
Does that alleviate the issue? We are consuming the extractor's container resources and would not put any risk of OOME to the process container, right?

	Os string `json:"os,omitempty"`
	// Hash of the version identifier of the Operating System (based on /etc/os-release VERSION_ID)
	OsVersion string `json:"osVersion,omitempty"`
	// Identifier of the kind of runtime
	Kind string `json:"kind,omitempty"`
	// Version of the kind of runtime
	KindVersion string `json:"kindVersion,omitempty"`
	// Entity that provides the runtime-kind implementation
	KindImplementer string `json:"kindImplementer,omitempty"`
	// Runtimes components
	Runtimes []RuntimeComponent `json:"runtimes,omitempty"`
}

type RuntimeComponent struct {
	// Name of a runtime used to run the application in the container
	Name string `json:"name,omitempty"`
	// The version of this runtime
	Version string `json:"version,omitempty"`
}
```

These new data elements are all optional (as they may not be relevant depending on the type of workloads) and are hashed (and truncated) to preserve shoulder-surfing anonymity (the hashing algorithm is [the same one that is currently used by the gatherer](https://github.com/openshift/insights-operator/blob/30b161f98eb185bb357ac84f8aca58e369d34bcd/pkg/gatherers/workloads/workloads_info.go#L336)).

> Someone with access to the date and who knows what they are looking for ("How many containers are running Go 1.21.3?") will be able to unobscure the data.

Reviewer: nit:
> Someone with access to the date...

I guess it should be `with the access to the data`

end
IOP->>IOP: call calculateWorkloadContainerShapes()
loop For each container
    IOP->>IOP: Poplulate the container shape with the hashed values of<br> the JSON in a runtimeInfo stanza if there is an entry<br> for the namespace/pod-name/container-id
end
IOP->>IOP: An archive is generated in /tmmp/insights-operator that contains <br>all gathered data (including the additional runtimeInfo for <br>the workload containers)

deactivate KRP
```

This is not a user-facing feature. It is achieved within the OpenShift Insights Operator gathering workflow.
The user is able to examine the gathered data by unarchiving the archive generated by the operator.

The DaemonSet for the insights-runtime-extractor is deployed when the Insights Operator is installed in the cluster and is not accessible outside of the `openshift-insights` namespace.

From the operator perspective, the insights-runtime-extractor is a black box, and its requirements are:

* Determine the Pod name of each insights-runtime-extractor’s Pod (as the insights-runtime-extractor is deployed as a DaemonSet, there is a single pod instance on each worker node) and map it to its worker node.

>   * There is no service exposing the insights-runtime-extractor and its pods are accessed directly throug their DNS records.

Reviewer: I think this is not true anymore. We have an exporter service.

Reviewer: Oh yeah right :+1: thanks

The extraction of data from the running container is achieved by executing “fingerprint” executables in the process `mnt` namespace to find the relevant information by reading specific files (e.g. the `Os` and `OsVersion` values are read from the `/etc/os-release` file if it is present). These fingerprints executables are self-contained executables.

This cannot be achieved by a program running in the OpenShift Insights Operator deployment.
This requires an additional resource for the operator, which is a DaemonSet that would deploy a “insights-runtime-extractor” image on all the cluster worker nodes.

The insights-runtime-extractor image requires high privileges to run (access the worker node’s host table, enter process namespaces, etc.) and needs to be configured with a specific security context constraint.

To minimize the code path running with high privileges, the insights-runtime-extractor is split into 2 components:

> * an "extractor" running in a privileged container. It runs a simple TCP server that, when it is trigged, will extract the runtime information and store them in a shared volume. The TCP server is bould to the loopback address so that it can only be contacted by clients running inside the same pod.

Reviewer: Typo in:
> The TCP server is bould

Splitting the inspection logic into multiple plugins/modules not only allows for extensibility but also provides for the ability to partition critical privilege requiring code from code which requires only point in time access to a specific container. Following this approach, privileged actions can be limited to a thin coordinator process in the `extractor` component, with the bulk of logic executed as a forked downgraded process limited to the same access afforded the inspected container target. Additionally, limiting the lifespan of the downgraded logic’s process to a per-container execution basis mitigates potential attacks where a rogue container corrupts the inspection logic to modify or gain access to a second victim container. 

##### Thin Privileged Coordinator Process

The risk of a workable exploit can be greatly reduced by minimizing the amount of total potential code, including the number and size of third-party dependencies. This comes from two key factors: First, the reduced complexity of the system limits the number of potential attack points, and secondly, the smaller total code set improves the speed and effectiveness of both human and automated code auditing.

> As such, three key key rules should be followed in the design and implementation of the coordinator process

Reviewer: Double `key` word

It would always be empty.

## Upgrade / Downgrade Strategy

Upgrades/downgrades of the insights-runtime-extractor will be handled by updating its image URL in the Insights Operator manifests.
The release cycle of the insights-runtime-extractor is bound to the Insight Operator release cycle.

This proposal does not bring additional change to the upgrade/downgrade of the OpenShift Insights Operator.

## Version Skew Strategy

This proposal does not impact the version skew strategy of the Insights Operator.

## Operational Aspects of API Extensions

The insights-runtime-extractor runs with its own cgroups and does not impact the inspected containers.

However, it will consume a certain amount of CPU and memory on each worker node while it is inspecting.
These consumptions must be bounded to avoid disrupting the user workload.

Inspecting containers will also increase the time to collect insights by the operator. The workload gatherer is performed every 12 hours and its execution will depend on the number of containers that are inspected.

---
**TBD** Provide measures of resource consumption (memory and CPU)

> 

Reviewer: Can we remove this or ?

Author: I removed it.
